src/json.c

```diff
@@ -77,7 +83,95 @@ int FieldSpec_CheckJsonType(FieldType fieldType, JSONType type) {
   return rv;
 }
 
-int JSON_StoreInDocField(RedisJSON json, JSONType jsonType, struct DocumentField *df) {
+// Uncomment when support for more types is added
+// static int JSON_getInt32(RedisJSON json, int32_t *val) {
+//   long long temp;
+//   int ret = japi->getInt(json, &temp);
+//   *val = (int32_t)temp;
+//   return ret;
+// }
+
+static int JSON_getFloat32(RedisJSON json, float *val) {
+  double temp;
+  int ret = japi->getDouble(json, &temp);
+  if (REDISMODULE_OK != ret) {
+    long long temp2;
+    ret = japi->getInt(json, &temp2);
+    *val = (float)temp2;
+  } else {
+    *val = (float)temp;
+  }
+  return ret;
+}
+
+// Uncomment when support for more types is added
+// static int JSON_getFloat64(RedisJSON json, double *val) {
+//   int ret = japi->getDouble(json, val);
+//   if (REDISMODULE_OK != ret) {
+//     long long temp;
+//     ret = japi->getInt(json, &temp);
+//     *val = (double)temp;
+//   }
+//   return ret;
+// }
+
+int JSON_StoreVectorInDocField(FieldSpec *fs, JSONResultsIterator arrIter, struct DocumentField *df) {
+  VecSimType type;
+  size_t dim;
+  int (*getElement)(RedisJSON, void *);
+
+  switch (fs->vectorOpts.vecSimParams.algo) {
+    case VecSimAlgo_HNSWLIB:
+      type = fs->vectorOpts.vecSimParams.hnswParams.type;
+      dim = fs->vectorOpts.vecSimParams.hnswParams.dim;
+      break;
+    case VecSimAlgo_BF:
+      type = fs->vectorOpts.vecSimParams.bfParams.type;
+      dim = fs->vectorOpts.vecSimParams.bfParams.dim;
+      break;
+    default: return REDISMODULE_ERR;
+  }
+  if (japi->len(arrIter) != dim) {
+    return REDISMODULE_ERR;
+  }
+
+  // The right function will put a value of the right type in the address given, or return REDISMODULE_ERR
+  switch (type) {
+    default:
+    case VecSimType_FLOAT32:
+      getElement = (int (*)(RedisJSON, void *))JSON_getFloat32;
+      break;
+    // Uncomment when support for more types is added
+    // case VecSimType_FLOAT64:
+    //   getElement = (int (*)(RedisJSON, void *))JSON_getFloat64;
+    //   break;
+    // case VecSimType_INT32:
+    //   getElement = (int (*)(RedisJSON, void *))JSON_getInt32;
+    //   break;
+    // case VecSimType_INT64:
+    //   getElement = (int (*)(RedisJSON, void *))japi->getInt;
+    //   break;
+  }
+
+  if (!(df->strval = rm_malloc(fs->vectorOpts.expBlobSize))) {
+    return REDISMODULE_ERR;
+  }
+  df->strlen = fs->vectorOpts.expBlobSize;
+
+  RedisJSON json;
+  unsigned char step = VecSimType_sizeof(type);
+  // At this point iterator length matches blob length
+  for (size_t offset = 0; (json = japi->next(arrIter)); offset += step) {
+    if (getElement(json, df->strval + offset) != REDISMODULE_OK) {
+      rm_free(df->strval);
+      return REDISMODULE_ERR;
+    }
+  }
+  df->unionType = FLD_VAR_T_CSTR;
```

Comment:
@DvirDukhan Wouldn't this require encoding/decoding, which we avoided by using blobs?

Reply:
I'll rephrase - here the blob is saved in the field `df->strval` and the type given to it is `FLD_VAR_T_CSTR`. My question is, since it is referred as a string value (with length), is to avoid someone will call `strdup` or `lower` or whatever string function on it.

Reply:
@DvirDukhan Not sure I understand the question. We have 2 field types to handle strings: `FLD_VAR_T_CSTR`, which we `rm_strndup` to keep the read-only value we get from `getString` API (see `JSON_StoreInDocField`) and the other option for a string field is `FLD_VAR_T_RMS`, which is using `RedisModuleString` allocated by `getJSON` API.

Reply:
As far as I saw we are always using the `df->strlen` value when referring to the string value, and the life span of this string is from this creation and until we pass it to `VecSimIndex_AddVector()` and then we free it